SI-5568 Fixes verify error from getClass on refinement of value type

JamesIry opened the pull request:
().asInstanceOf[AnyRef with Unit].getClass and
5.asInstanceOf[AnyRef with Int].getClass would cause a verify
error. Going the other way, i.e. [Unit with AnyRef] or [Int with AnyRef]
worked fine. This commit fixes it that both directions work out to
BoxedUnit or java.lang.Integer.

review @gkossakowski

JamesIry committed:
SI-5568 Fixes verify error from getClass on refinement of value type

().asInstanceOf[AnyRef with Unit].getClass and
5.asInstanceOf[AnyRef with Int].getClass would cause a verify
error. Going the other way, i.e. [Unit with AnyRef] or [Int with AnyRef]
worked fine. This commit fixes it that both directions work out to
BoxedUnit or java.lang.Integer.

gkossakowski commented on src/compiler/scala/tools/nsc/transform/Erasure.scala:
```diff
@@ -1152,9 +1158,14 @@ abstract class Erasure extends AddInterfaces
+              } else if (primitiveGetClassMethods.contains(fn.symbol)) {
+                // if we got here then we're trying to send a primitive getClass method to 
+                // a) an Any, or
+                // b) a non-primitive, e.g. because the qualifier's type is a refinement type where part of the refinement is a primitive.
```
`part` is probably not the best word. Did you mean `parent`?

JamesIry replied:
parent is fine

retronym commented on src/compiler/scala/tools/nsc/transform/Erasure.scala:
```diff
@@ -340,12 +340,18 @@ abstract class Erasure extends AddInterfaces
-  private lazy val interceptedMethods = Set[Symbol](Any_##, Object_##, Any_getClass, AnyVal_getClass) ++ (
-    // Each value class has its own getClass for ultra-precise class object typing.
+  
+  // Each value class has its own getClass for ultra-precise class object typing.
```
I guess we need a more precise term for "value class" now that `AnyVal` is no longer sealed. "Primitive value class"?

JamesIry replied:
I'm good with that

gkossakowski commented on src/compiler/scala/tools/nsc/transform/Erasure.scala:
```diff
@@ -340,12 +340,18 @@ abstract class Erasure extends AddInterfaces
+  
+  // Methods on Any/Object which we rewrite here while we still know what
+  // is a primitive and what arrived boxed.
+  private lazy val interceptedMethods = poundPoundMethods ++ primitiveGetClassMethods
```
Thanks for cleaning this up!

gkossakowski commented:
@JamesIry, it would be useful to add a comment in commit message why boxing is now properly triggered (it's handled later in erasure phase). I know you mention that on a ticket but it's easier to use tools like git blame if essential info (conclusion) is in commit message. Please do that only if you are about to change something else (e.g change comments in code).

Other than that LGTM!

JamesIry commented:
I can beef up that comment while I'm working on the other comment bits.

JamesIry committed:
SI-5568 Comment improvements for getClass on primitive intersection.

Based on code review here are a few comment cleanups and the removal of
some dead test code.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1418/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2144/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1418/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2144/

JamesIry commented:
ping @gkossakowski for final review

gkossakowski commented:
@JamesIry: I didn't notice your new commit. Everything Looks Good to Me now.
